libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -899,3 +952,435 @@ float AP_GPS::get_lag(uint8_t instance) const
         return drivers[instance]->get_lag();
     }
 }
+
+/*
+ calculate the weightings used to blend GPs location and velocity data
+*/
+void AP_GPS::calc_blend_weights(void)
+{
+    // zero the blend weights
+    memset(&_blend_weights, 0, sizeof(_blend_weights));
+
+    // determine how many GPS receivers are returning a fix and the instance of the first receiver that is
+    uint8_t first_instance = 0;
+    for (uint8_t i=0; i<GPS_MAX_INSTANCES; i++) {
+        if (state[i].status > NO_FIX) {
+            num_instances = i+1;
+            if (first_instance == 0) {
+                first_instance = i;
+            }
+        }
+    }
+
+    // If not enough receivers to do blending or blending has  been disabled, set the primary instance and weighting to the first receiver detected
+    if (num_instances < 2 || _auto_switch != 2) {
+        primary_instance = first_instance;
+        _output_is_blended = false;
+        _blend_weights[primary_instance] = 1.0f;
+        return;
+    }
+
+    // Use the oldest non-zero time, but if time difference is excessive, use newest to prevent a disconnected receiver from blocking updates
+    uint32_t max_ms = 0; // newest non-zero system time of arrival of a GPS message
+    uint32_t min_ms = -1; // oldest non-zero system time of arrival of a GPS message
+    uint8_t newest_instance = 0; // index of receiver with newest time stamp
+    int16_t max_rate_ms = 0; // largest update interval of a GPS receiver
+    for (uint8_t i=0; i<GPS_MAX_INSTANCES; i++) {
+        // Find largest and smallest times
+        if (state[i].last_gps_time_ms > max_ms) {
+            max_ms = state[i].last_gps_time_ms;
+            newest_instance = i;
+        }
+        if ((state[i].last_gps_time_ms < min_ms) && (state[i].last_gps_time_ms > 0)) {
+            min_ms = state[i].last_gps_time_ms;
+        }
+        if (_rate_ms[i] > max_rate_ms) {
+            max_rate_ms = _rate_ms[i];
```

Comment:
It does matter if the GPS is slower than the parameter - eg the parameter was set to 100ms and the GPs was updating at 200. It would likely result in the data from that GPS being ignored by the blending algorithm due to failing the data delayed check at line 1001.

Given that the default parameter value is 200msec, I don't think this is a problem as we would not want to use a GPS receiver that outputted slower than that anyway. If a user sets the parameter to a different value, then they are responsible for checking that their GPS is capable of that rate.